### DISCUSSION TITLE
Need help debugging a policy matching behavior

### DISCUSSION THREAD

**Discussion by bsiara** [rabbitmq/rabbitmq-server#11878] [Other] [answered]
### Describe the bug

I have 3 nodes rmq cluster, version 3.13.6. On this cluster I have 2 policies which match error or non error queues:
- ha-all
```
Pattern 	^(.*(?<![E|e]rror))?$
Apply to 	all
Definition 	
federation-upstream-set:	all
ha-mode:	all
ha-promote-on-failure:	always
ha-promote-on-shutdown:	always
ha-sync-mode:	automatic
message-ttl:	2592000000
Priority 	0
```
- ha-all-error
```
Pattern 	^.*[E|e]rror$
Apply to 	all
Definition 	
federation-upstream-set:	all
ha-mode:	all
ha-promote-on-failure:	always
ha-promote-on-shutdown:	always
ha-sync-mode:	automatic
message-ttl:	60000
Priority 	0
```

Policy ha-all-error did not work properly and did not match all queues.
![policies](https://github.com/user-attachments/assets/74651601-ea76-4e1c-af1e-b4fee531b15f)
![queues](https://github.com/user-attachments/assets/c183a53e-863c-4093-ba49-57e34c97a035)

### Reproduction steps

1. Create policies as described
- ha-all
```
Pattern 	^(.*(?<![E|e]rror))?$
Apply to 	all
Definition 	
federation-upstream-set:	all
ha-mode:	all
ha-promote-on-failure:	always
ha-promote-on-shutdown:	always
ha-sync-mode:	automatic
message-ttl:	2592000000
Priority 	0
```
- ha-all-error
```
Pattern 	^.*[E|e]rror$
Apply to 	all
Definition 	
federation-upstream-set:	all
ha-mode:	all
ha-promote-on-failure:	always
ha-promote-on-shutdown:	always
ha-sync-mode:	automatic
message-ttl:	60000
Priority 	0
```
3. Create queues with different names and with `(E|e)rror` in name at the end

### Expected behavior

Policy match all queues as UI matched.

**Comment by michaelklishin** [comment, marked as answer]
@bsiara please never use issues for questions in repositories that have them enabled.

If policy pattern matching was "broken" we would be flooded with such questions but that's not the case.

For starters, note that you have two policies with [identical priorities](https://www.rabbitmq.com/docs/parameters#policy-priorities). In which case RabbitMQ will randomly select the applicable one.

On an unrelated note, your settings suggest that you use [classic mirrored queues](https://www.rabbitmq.com/docs/ha), which have been **removed** from RabbitMQ for 4.x after three years of deprecation. Please [move to quorum queues](https://www.rabbitmq.com/docs/quorum-queues) and/or non-replicated classic queues, both of which can be federated.

**Comment by bsiara** [comment]
Hi Michael, thanks for your answer. If you look a bit deeper to regexp in this policies first match all queues without 'error' in name at the end and second match all queues which include 'error' in name at the and. In this case priority does not matter because both policies did not match the same queue, but maybe you are right I changed priority on this policies and policy ha-all-error still did not work and did not match all queues with 'error' at the end.
![Zrzut ekranu z 2024-08-01 17-43-51](https://github.com/user-attachments/assets/c75e7142-4adf-4a92-9657-f1ad18287464)
![Zrzut ekranu z 2024-08-01 17-44-05](https://github.com/user-attachments/assets/fc0c258e-df7e-4cac-9414-837398627820)

**Reply by michaelklishin** [reply]
`^(.*(?<![E|e]rror))?$` does not match `abc.error` with three regular expression implementations I have tried quickly: Erlang, Python, Ruby.

I'm afraid our team won't troubleshoot regular expressions for you.

**Reply by michaelklishin** [reply]
According to https://regexr.com/, the problem could come down to the use of negative lookbehind, a feature that not every regular expressions engine supports, and not every one that supports it enables by default without additional flags.

Erlang's implementation [does support them](https://www.erlang.org/doc/apps/stdlib/re.html), see the `Lookbehind Assertions` section. I could not find any evidence that it would require the use of additional "extended" flags.

I am out of other hypothesis but it does not look like there's anything obviously missing in RabbitMQ itself or its runtime's regular expressions engine.

I have pointed out that your example uses two features/configuration settings it should not: classic mirrored queues and conflicting policy priorities. Plus you apply your policies to everything, which will further increase the risk of conflicts and confusing matching behavior.

**Reply by bsiara** [reply]
You try wrong regexp, ha-all-error has definition:
```
^.*[E|e]rror$
```
and this definition did not use Lookbehind Assertions. When you use this definition this match `abc.error` name. Rabbitmq management ui match this queues properly but policies regex did not match all queues.
![Zrzut ekranu z 2024-08-01 19-17-21](https://github.com/user-attachments/assets/3d4f2bc1-da0f-4948-8ad6-43715c2ce0d6)
![Zrzut ekranu z 2024-08-01 19-28-00](https://github.com/user-attachments/assets/b53ffdcb-082b-4ddd-afb7-c3a43db31572)

**Comment by lukebakken** [comment]
You should simplify your regex. `[]` is a character class so using `|` is not necessary. `[E|e]` is a character class that matches `E`, `|`, and `e`, which isn't what you want.

`^.*[E|e]rror$` should be `\.[Ee]rror$`

As for the other regex, you should make similar simplifications.

**Reply by bsiara** [reply]
@lukebakken I made corects as you suggest but still did not help, policy ha-all-error do not match all queues with 'error' at the end but ui make this
